#                                            IMBD Analysis for Learning NLP basics

We are going to use Bag of words for pre-processing the text and apply different classification algorithms to classify whether a given movie review is **positive** or **negative**. 


## About Data : IMBD Dataset

Credits: https://www.kaggle.com/datasets/lakshmi25npathi/imdb-dataset-of-50k-movie-reviews?resource=download

> - This data consists of two columns. - review - sentiment
> - Reviews are the statements given by users after watching the movie.
> - sentiment feature tells whether the given review is positive or negative.

In [1]:
# Import  necessary librairies
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.ensemble import RandomForestClassifier
from  sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline
from sklearn.metrics import classification_report

In [4]:
df = pd.read_csv("IMDB Dataset.csv")
df.shape

(50000, 2)

In [5]:
df.head(5)

review sentiment
0  One of the other reviewers has mentioned that ...  positive
1  A wonderful little production. <br /><br />The...  positive
2  I thought this was a wonderful way to spend ti...  positive
3  Basically there's a family where a little boy ...  negative
4  Petter Mattei's "Love in the Time of Money" is...  positive

In [ ]:
#creating a new column "Category" which represent 1 if the sentiment is positive or 0 if it is negative
df['Category']=df.sentiment.apply(lambda x : 1 if x=="positive" else 0)
df

review sentiment  Category
0      One of the other reviewers has mentioned that ...  positive         1
1      A wonderful little production. <br /><br />The...  positive         1
2      I thought this was a wonderful way to spend ti...  positive         1
3      Basically there's a family where a little boy ...  negative         0
4      Petter Mattei's "Love in the Time of Money" is...  positive         1
...                                                  ...       ...       ...
49995  I thought this movie did a down right good job...  positive         1
49996  Bad plot, bad dialogue, bad acting, idiotic di...  negative         0
49997  I am a Catholic taught in parochial elementary...  negative         0
49998  I'm going to have to disagree with the previou...  negative         0
49999  No one expects the Star Trek movies to be high...  negative         0

[50000 rows x 3 columns]

In [9]:
##check the distribution of 'Category' and see whether the Target labels are balanced or not.
df.Category.value_counts()

Category
1    25000
0    25000
Name: count, dtype: int64

In [10]:
X_train,X_test,y_train,y_test=train_test_split(df.review,df.Category,test_size=0.2)

In [14]:
#using sklearn pipeline module to create a classification pipeline to classify the movie review with KNN
clf1=Pipeline([('cv',CountVectorizer()),('knn',KNeighborsClassifier(n_neighbors=10,metric='euclidean'))])

In [15]:
clf1.fit(X_train,y_train)

Pipeline(steps=[('cv', CountVectorizer()),
                ('knn',
                 KNeighborsClassifier(metric='euclidean', n_neighbors=10))])

In [17]:
y_pred=clf1.predict(X_test)
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.65      0.65      0.65      4982
           1       0.65      0.65      0.65      5018

    accuracy                           0.65     10000
   macro avg       0.65      0.65      0.65     10000
weighted avg       0.65      0.65      0.65     10000



In [18]:
#using sklearn pipeline module to create a classification pipeline to classify the movie review with Multinomial Naive Bayes
clf2=Pipeline([('cv',CountVectorizer()),('MNB',MultinomialNB())])

In [19]:
clf2.fit(X_train,y_train)

Pipeline(steps=[('cv', CountVectorizer()), ('MNB', MultinomialNB())])

In [20]:
y_pred=clf2.predict(X_test)
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.83      0.88      0.85      4982
           1       0.87      0.82      0.84      5018

    accuracy                           0.85     10000
   macro avg       0.85      0.85      0.85     10000
weighted avg       0.85      0.85      0.85     10000



In [22]:
#using sklearn pipeline module to create a classification pipeline to classify the movie review with  Random Forest
clf3=Pipeline([('cv',CountVectorizer()),('RFC',RandomForestClassifier(n_estimators=50,criterion='entropy'))])

In [23]:
clf3.fit(X_train,y_train)

Pipeline(steps=[('cv', CountVectorizer()),
                ('RFC',
                 RandomForestClassifier(criterion='entropy', n_estimators=50))])

In [24]:
y_pred=clf3.predict(X_test)
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.84      0.84      0.84      4982
           1       0.84      0.84      0.84      5018

    accuracy                           0.84     10000
   macro avg       0.84      0.84      0.84     10000
weighted avg       0.84      0.84      0.84     10000



### Summary for this Analysis

**Quick Comparison of results**

| Classifier | Precision | recall | f1-score |
| ----------- | ----------- | ----------- | ----------- |
| KNN | 0.65 | 0.65 | 0.65 |
| MultinomialNB | 0.85 | 0.85 | 0.845|
| RandomForest | 0.84 | 0.84 | 0.84 |


> - As Machine learning algorithms does not work on Text data directly, we need to convert them into numeric vector and feed that into models while training.

>- In this process, we convert text into a very high dimensional numeric vector using the technique of Bag of words.
Model like K-Nearest Neighbours(KNN) doesn't work well with high dimensional data because with large number of dimensions, it becomes difficult for the algorithm to calculate distance in each dimension. In higher dimensional space, the cost to calculate distance becomes expensive and hence impacts the performance of model.

> - The easy calculation of probabilities for the words in corpus(Bag of words) and storing them in contigency table is the major reason for the Multinomial NaiveBayes to be a text classification friendly algorithm.

> - As Random Forest uses Bootstrapping(Row and column Sampling) with many decision tree and overcomes the high variance and overfitting of high dimensional data and also uses feature importance of words for better classifing the categories.

> - Machine Learning is like trial and error scientific method, where we keep trying all the possible algorithms we have and select the one which give good results and satisfy the requirements like latency, interpretability etc.